File: importer/vigilo/download.py
#!/usr/bin/python
import os
import re
import sys
import json
#import csv
import glob
#import argparse
import zipfile
import pandas as pd
import numpy as np
import requests
import glob

import sqlite3 as sql
from packaging import version

# Add module path
sys.path.insert(0, os.path.abspath(
    os.path.join(os.path.dirname(__file__), "../..")))

import lib.python.net as net
import lib.python.file as file

# # Set pandas options
pd.set_option('mode.chained_assignment', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', 1000)

DEFAULT_MISSING = pd._libs.parsers.STR_NA_VALUES
DEFAULT_MISSING = DEFAULT_MISSING.remove('NA')

# Create temorary folder
if not os.path.exists('./downloaded/vigilo'):
    os.makedirs('./downloaded/vigilo')

def download_instances():
    df = pd.read_json('https://vigilo-bf7f2.firebaseio.com/citylist.json',orient='index')
    df.reset_index(inplace=True)
    df.rename(columns={'index': 'name'},inplace=True)
    df=df[df['prod']]
    df.drop(["prod"], axis=1, inplace=True)
    df.insert(0,'InstanceID',df.index + 1) 
    df.to_csv('./downloaded/vigilo/instance.csv',index=False)

    return df

def download_categories():
    df = pd.read_json('https://vigilo-bf7f2.firebaseio.com/categorieslist.json',orient='values')
    df.to_csv('./downloaded/vigilo/category.csv',index=False)


def download_observations(instances):
    for index, row in instances.iterrows():
        api_path = row['api_path'] 
        scope = row['scope'] 
        instance_id = row['InstanceID']
        try:
            # Get version
            r = requests.get(f'{api_path}/get_version.php', allow_redirects=True)
            apiversion = r.json()['version']
            if version.parse(apiversion) > version.parse("0.0.11"):
                df = pd.read_json(f"{api_path}/get_issues.php?scope={scope}&format=json")
                df.insert(0,'InstanceID',instance_id)
                df.to_csv(f'./downloaded/vigilo/obs_{scope}.csv',index=False)
                print(f"Downloader {scope} => {api_path}")
            else:
                print(f"!! BAD API VERSION !! {scope} => {api_path}")
        except:
            print(f"!! ERROR !! {scope} => {api_path}")
            pass

def concatenate_observation():
    modes = ['w','a']

    files = glob.glob('./downloaded/vigilo/obs_*.csv')
    for idx in range(len(files)):
        filename = files[idx]
        mode = modes[min(idx,1)]

        df = pd.read_csv(filename)
        df.to_csv('./downloaded/vigilo/observation.csv',mode=mode,index=False,header=idx<1)

#######################################
# Main
#######################################

download_categories()
instances = download_instances()
download_observations(instances)
concatenate_observation()

File: importer/vigilo/download.R
suppressPackageStartupMessages({
  library(rio)
  library(httr)
  library(dplyr)
  library(RSQLite)
  library(jsonlite)
})

# Commons vars
dlfolder <- "./downloaded/vigilo/"
dbfolder <- "./db/"

# Make export directory
dir.create(dlfolder,recursive=TRUE, showWarnings = FALSE)

################################################################################
# Function
################################################################################

# Get categories
download_categories <- function() {
  print("Try to download categories")

  url <- "https://vigilo-bf7f2.firebaseio.com/categorieslist.json"
  jcategories <- jsonlite::fromJSON(url)

  # Format
  jcategories <- jcategories %>%
    mutate_at(vars("catdisable"), ~replace(., is.na(.), 0)) 
  #jcategories$catdisable <- as.logical(jcategories$catdisable)

  # Export CSV
  rio::export(jcategories, paste0(dlfolder, "category.csv"))
}

# Get all instances
download_instances <- function() {
  # Get instances
  print("Try to download instances informations")

  url <- "https://vigilo-bf7f2.firebaseio.com/citylist.json"
  file_content <- paste("[", readLines(url, warn = FALSE), "]")
  jinstances <- jsonlite::fromJSON(file_content, simplifyVector = FALSE)
  df <- Reduce(rbind, jinstances[[1]])

  scopenames <- names(jinstances[[1]])
  instances <- data.frame(scopename = scopenames, df, row.names = NULL)

  instances <- instances %>%
    filter(prod == TRUE)
  nb_instances <- nrow(instances)

  # Get only compatible version> 0.0.11
  versions <- c(rep("0.0.0"),nb_instances)
  compatibles <- c(rep(FALSE,nb_instances))
  for (idx in seq_len(nrow(instances)))
  {
    tryCatch(
      {
        print(paste("Try get ",instances$scope[idx],"instance version"))

        url <- paste0(instances$api_path[idx], "/get_version.php")
        jcontent <- jsonlite::fromJSON(url)
        versions[idx] <- jcontent$version
        compatibles[idx] <- compareVersion("0.0.11",jcontent$version) == -1
      },
      error = function(cond) {
      }
    )
  }

  instances$version <- versions  
  instances$compatible <- compatibles  

  instances <- instances %>%
    filter(compatible==TRUE)
  nb_instances <- nrow(instances)

  # Get limitbox
  # "coordinate_lat_min": "50.371981",
  # "coordinate_lat_max": "50.526368",
  # "coordinate_lon_min": "3.837876",
  # "coordinate_lon_max": "4.153979"
  instance_isup <- c(rep(FALSE,nb_instances))
  lat_mins <- c(rep(0,nb_instances))  
  lat_maxs <- c(rep(0,nb_instances))
  lon_mins <- c(rep(0,nb_instances))
  lon_maxs <- c(rep(0,nb_instances))

  for (idx in seq_len(nrow(instances)))
  {
    tryCatch(
      {
        print(paste("Try to download",instances$scope[idx],"instance observations"))

        url <- paste0(instances$api_path[idx], "/get_scope.php?scope=",instances$scope[idx])
        jcontent <- jsonlite::fromJSON(url)

        lat_mins[idx] <- jcontent$coordinate_lat_min
        lat_maxs[idx] <- jcontent$coordinate_lat_max
        lon_mins[idx] <- jcontent$coordinate_lon_min
        lon_maxs[idx] <- jcontent$coordinate_lon_max
        instance_isup <- TRUE
      },
      error = function(cond) {
      }
    )
  }

  # Filter compatible versions
  
  instances$isup <- instance_isup  
  instances$lat_min <- lat_mins
  instances$lat_max <- lat_maxs
  instances$lon_min <- lon_mins
  instances$lon_max <- lon_maxs
  instances <- instances %>%
    filter(isup)


  # Remove prod column and add scopid column
  instances$scopeid <- seq_len(nrow(instances))
  
  instances <- instances %>%
    subset(select=-c(prod,isup,compatible)) %>%
    relocate("scopeid")

  # Export CSV
  rio::export(instances, paste0(dlfolder, "instance.csv"))

  return(instances)
}

# Download observations
download_observations <- function(instances) {
  for (idx in seq_len(nrow(instances))) {
    filename <- paste0(dlfolder, "/obs_", instances$scope[idx], ".csv")
    url <- paste0(
      instances$api_path[idx],
      "/get_issues.php?scope=",
      instances$scope[idx],
      "&format=json"
    )

    # Add instanceid column
    observations <- jsonlite::fromJSON(url)

    # If no observations, continue
    if (length(observations)==0) next

    # Save observations
    observations$instanceid <- idx 
    observations <- observations %>%
      relocate("instanceid")

    rio::export(observations, filename)
  }
}

concatenate_observations <- function() {
  observations <- list.files(
    path = "downloaded/vigilo/",
    pattern = "obs_",
    full.name = TRUE
  ) %>%
    lapply(rio::import) %>%
    bind_rows()

  rio::export(observations, paste0(dlfolder, "observation.csv"))
}

################################################################################
# Main
################################################################################

options(timeout = 10)

download_categories()
instances <- download_instances()
download_observations(instances)
concatenate_observations()